lnrpc/routerrpc/router_server.go

```diff
@@ -747,3 +752,244 @@ func (s *Server) BuildRoute(ctx context.Context,
 
 	return routeResp, nil
 }
+
+// SubscribeHtlcEvents creates a uni-directional stream from the server to
+// the client which delivers a stream of htlc events.
+func (s *Server) SubscribeHtlcEvents(req *SubscribeHtlcEventsRequest,
+	stream Router_SubscribeHtlcEventsServer) error {
+
+	htlcClient, err := s.cfg.RouterBackend.SubscribeHtlcEvents()
+	if err != nil {
+		return err
+	}
+	defer htlcClient.Cancel()
+
+	for {
+		select {
+		case event := <-htlcClient.Updates():
+			htlcEvent, ok := event.(htlcswitch.HtlcEvent)
+			if !ok {
+				return fmt.Errorf("unexpected event type: %T",
+					event)
+			}
+
+			rpcEvent, err := rpcHtlcEvent(htlcEvent)
+			if err != nil {
+				return err
+			}
+
+			switch e := event.(type) {
+			case *htlcswitch.ForwardingEvent:
+				rpcEvent.Event = &HtlcEvent_ForwardEvent{
+					ForwardEvent: &ForwardEvent{
+						HtlcInfo: rpcHTLCInfo(e.HtlcInfo),
+					},
+				}
+
+			case *htlcswitch.ForwardingFailEvent:
+				rpcEvent.Event = &HtlcEvent_ForwardFailEvent{
+					ForwardFailEvent: &ForwardFailEvent{},
+				}
+
+			case *htlcswitch.LinkFailEvent:
+				failureCode, failReason, err := rpcFailReason(
+					e.LinkError,
+				)
+				if err != nil {
+					return err
+				}
+
+				rpcEvent.Event = &HtlcEvent_LinkFailEvent{
+					LinkFailEvent: &LinkFailEvent{
+						HtlcInfo:      rpcHTLCInfo(e.HtlcInfo),
+						WireFailure:   failureCode,
+						FailureDetail: failReason,
+						FailureString: e.LinkError.Error(),
+					},
+				}
+
+			case *htlcswitch.SettleEvent:
+				rpcEvent.Event = &HtlcEvent_SettleEvent{
+					SettleEvent: &SettleEvent{},
+				}
+			}
+
+			if err := stream.Send(rpcEvent); err != nil {
+				return err
+			}
+
+		case <-s.quit:
+			return errors.New("server shutting down")
+		}
+	}
+}
+
+// rpcHtlcEvent returns a rpc htlc event. Note that the event field is not set
+// by this function, only the fields that are common to all events are set.
+func rpcHtlcEvent(htlcEvent htlcswitch.HtlcEvent) (*HtlcEvent, error) {
```

Comment:
Move some of this code to a new file, to not go down the way that `rpcserver.go` went? (very long file)